rsocket-core/src/main/java/io/rsocket/buffer/Tuple2ByteBuf.java

```diff
@@ -0,0 +1,415 @@
+package io.rsocket.buffer;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.util.ReferenceCountUtil;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.charset.Charset;
+import org.agrona.BufferUtil;
+
+class Tuple2ByteBuf extends AbstractTupleByteBuf {
+  private static final int MEMORY_CACHE_ALIGNMENT = 64;
+
+  private static final long ONE_MASK = 0x100000000L;
+  private static final long TWO_MASK = 0x200000000L;
+  private static final long MASK = 0x700000000L;
+  private static final boolean CHECK_BOUNDS = true;
+  private static final ByteBuffer EMPTY_NIO_BUFFER = Unpooled.EMPTY_BUFFER.nioBuffer();
+
+  private final ByteBuf one;
+  private final ByteBuf two;
+  private final int oneReadIndex;
+  private final int twoReadIndex;
+  private final int oneReadableBytes;
+  private final int twoReadableBytes;
+  private final int twoRelativeIndex;
+
+  private boolean freed;
+
+  Tuple2ByteBuf(ByteBufAllocator allocator, ByteBuf one, ByteBuf two) {
+    super(allocator, one.readableBytes() + two.readableBytes());
+
+    this.one = one;
+    this.two = two;
+
+    this.oneReadIndex = one.readerIndex();
+    this.twoReadIndex = two.readerIndex();
+
+    this.oneReadableBytes = one.readableBytes();
+    this.twoReadableBytes = two.readableBytes();
+
+    this.twoRelativeIndex = oneReadableBytes;
+
+    this.freed = false;
+  }
+
+  public long calculateRelativeIndex(int index) {
+    if (CHECK_BOUNDS && index > capacity) {
```

Comment:
We should be able to disable bounds checking - this should be a property. I believe there could be a property netty uses for its bounds checking - we might want to use use the same one.

Reply:
rigth

Reply:
fixed